File: public/packages/typescript/geo/abstractions/src/name-resolution/name-normalizer.ts
// -----------------------------------------------------------------------
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) DCSV
// -----------------------------------------------------------------------

import { falsey } from "@d2/utilities";

/**
 * Name-normalization pipeline used by the geo name-resolver fuzzy matcher
 * to fold user-supplied free-form text down to a canonical comparison
 * form. Pure function — no I/O, no shared state, thread safe by design.
 *
 * Mirrors .NET `D2.Shared.Geo.Abstractions.NameResolution.NameNormalizer`
 * byte-for-byte; the same input MUST produce the same output on both
 * runtimes. Cross-language parity is pinned via the byte-equivalent
 * fixture exercised by the parity test suite.
 *
 * Pipeline (applied in the same order as .NET — step ordering is
 * load-bearing: future engineers adding case-sensitive transforms must
 * NOT reorder these without auditing both runtimes):
 *
 * 1. **Empty / whitespace-only short-circuit** — returns `""` so callers
 *    can pipe raw user input in without first guarding against blanks.
 *    Implemented via `falsey()` from `@d2/utilities` (the parity helper
 *    that mirrors the .NET `Falsey()` extension semantics — null /
 *    undefined / empty / whitespace-only all collapse to `true`). The
 *    function signature rejects `null` / `undefined` at compile time, so
 *    the runtime null branches are unreachable from typed callers; the
 *    helper still handles the empty + whitespace-only cases.
 * 2. **NFD decomposition** — splits accented characters into base letter
 *    + combining mark so the marks can be stripped in the next pass.
 * 3. **Strip combining marks** — `\p{M}` Unicode property removes every
 *    diacritic / accent (`"São Paulo"` → `"Sao Paulo"`).
 * 4. **Spaced ampersand-token substitution** — replaces the spaced form
 *    `" & "` (with whitespace on BOTH sides) with `" and "` so
 *    `"Trinidad & Tobago"` and `"Trinidad and Tobago"` compare equal.
 *    The spaced-form requirement deliberately preserves unspaced
 *    ampersands like `"AT&T"`, `"M&M's"`, `"R&D"` — applied BEFORE
 *    casefolding to match the .NET pipeline order exactly (the substring
 *    is case-invariant so the output is byte-equivalent either way, but
 *    order alignment is enforced so future case-sensitive steps cannot
 *    silently drift).
 * 5. **Locale-invariant lower-case** — `String.prototype.toLowerCase()`
 *    without a locale argument is ECMAScript locale-independent (matches
 *    .NET `CultureInfo.InvariantCulture.TextInfo.ToLower`). Note: a
 *    handful of edge-case Unicode code points (e.g. Turkish dotted-I
 *    `İ`) fold differently between ECMAScript's default case-fold and
 *    .NET's `ToLowerInvariant`. These are exceedingly rare in geo
 *    reference data and the cross-language parity fixture is the
 *    trip-wire that flags any real drift.
 * 6. **Trim outer whitespace** — `.trim()` removes leading / trailing
 *    whitespace.
 * 7. **Collapse internal whitespace** — runs of one or more whitespace
 *    characters reduce to a single ASCII space (`"United  States"` →
 *    `"United States"`).
 *
 * Pure (no closure state, no module-level mutation) — safe to call from
 * any concurrent caller.
 *
 * @param input - the free-form text to normalize. Empty / whitespace-only
 *   input returns `""`. Callers that need to distinguish "no input" from
 *   "valid-but-not-found" should check `=== ""` (or `falsey()` from
 *   `@d2/utilities`) after the call.
 * @returns the normalized comparison key, or `""` for empty /
 *   whitespace-only input.
 */
export function normalize(input: string): string {
  // 1. Empty / whitespace-only short-circuit via the shared `falsey()`
  //    helper from `@d2/utilities` — mirrors the .NET `input.Falsey()`
  //    extension semantics so both runtimes collapse the same set of
  //    "no signal" inputs (null / undefined / empty / whitespace-only).
  if (falsey(input)) return "";

  // 2. NFD decomposition.
  // 3. Strip combining marks (Unicode property \p{M}).
  const stripped = input.normalize("NFD").replace(/\p{M}/gu, "");

  // 4. Spaced ampersand-token substitution — spaced form only, preserves
  //    AT&T / M&M's / R&D / etc. Applied BEFORE casefold to match the
  //    .NET pipeline order exactly.
  const ampersandSwapped = stripped.replaceAll(" & ", " and ");

  // 5. Locale-invariant lowercase (matches .NET ToLowerInvariant — see
  //    TSDoc above for the Turkish dotted-I caveat).
  const lowered = ampersandSwapped.toLowerCase();

  // 6. Trim outer whitespace.
  // 7. Collapse internal whitespace runs to a single ASCII space.
  return lowered.trim().replace(/\s+/g, " ");
}

File: public/packages/dotnet/geo/abstractions/NameResolution/NameNormalizer.cs
// -----------------------------------------------------------------------
// <copyright file="NameNormalizer.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Geo.Abstractions.NameResolution;

using System.Globalization;
using System.Text;
using DcsvIo.D2.Utilities.Extensions;

/// <summary>
/// Pure helper that normalizes free-form place names (countries,
/// subdivisions, etc.) into a canonical comparison form before they are
/// fed into the geo name resolver's exact / fuzzy lookup table. The
/// algorithm is deterministic, stateless, and thread-safe — concurrent
/// callers can share a single reference to <see cref="Normalize"/>.
/// </summary>
/// <remarks>
/// <para>
/// <b>Cross-language parity:</b> the TypeScript mirror in
/// <c>@dcsv-io/d2-geo-abstractions</c> (<c>src/name-resolution/name-normalizer.ts</c>)
/// implements the byte-equivalent algorithm using
/// <c>String.prototype.normalize("NFD")</c> + a combining-mark strip +
/// <c>toLocaleLowerCase("en-US")</c>. The two implementations MUST
/// produce the same output for the same input — divergence breaks the
/// name-resolver fixtures the cross-language parity suite enforces.
/// </para>
/// <para>
/// <b>Pipeline order (each step is observable on its own):</b>
/// </para>
/// <list type="number">
///   <item>
///     <description>Null / empty / whitespace input short-circuits to
///     <see cref="string.Empty"/> so callers can pipe raw user input in
///     without first guarding against <c>null</c>.</description>
///   </item>
///   <item>
///     <description>Unicode NFD decomposition (<see cref="NormalizationForm.FormD"/>) —
///     splits accented characters into base letter + combining mark so the
///     marks can be stripped in the next pass.</description>
///   </item>
///   <item>
///     <description>Strip all combining marks
///     (<see cref="UnicodeCategory.NonSpacingMark"/>) so "São Paulo" and
///     "Sao Paulo" compare equal.</description>
///   </item>
///   <item>
///     <description>Replace the standalone ampersand token (<c> &amp; </c>
///     with surrounding whitespace) with <c> and </c> so "Trinidad &amp;
///     Tobago" and "Trinidad and Tobago" compare equal. We deliberately
///     match the spaced form only — touching every <c>&amp;</c> inside a
///     name would mangle things like <c>"AT&amp;T"</c>.</description>
///   </item>
///   <item>
///     <description>Casefold via
///     <see cref="CultureInfo.InvariantCulture"/>'s <see cref="TextInfo.ToLower(string)"/>
///     — invariant culture avoids the Turkish "dotless i" pitfall where
///     <c>"I".ToLower("tr-TR")</c> returns <c>"ı"</c>.</description>
///   </item>
///   <item>
///     <description>Trim leading / trailing whitespace.</description>
///   </item>
///   <item>
///     <description>Collapse runs of internal whitespace
///     (<see cref="char.IsWhiteSpace(char)"/>) down to a single ASCII
///     space so "United  States" and "United States" compare equal.</description>
///   </item>
/// </list>
/// </remarks>
public static class NameNormalizer
{
    private const char _SPACE = ' ';
    private const string _AMPERSAND_TOKEN = " & ";
    private const string _AND_TOKEN = " and ";

    /// <summary>
    /// Returns the canonical comparison form of <paramref name="input"/>.
    /// Returns <see cref="string.Empty"/> when <paramref name="input"/> is
    /// <c>null</c>, empty, or whitespace-only. Pure / stateless / threadsafe.
    /// </summary>
    /// <param name="input">
    /// The raw place name to normalize. May be <c>null</c>, empty, or
    /// whitespace-only — the helper never throws on bad input.
    /// </param>
    /// <returns>The canonical comparison form.</returns>
    public static string Normalize(string? input)
    {
        if (input.Falsey())
            return string.Empty;

        // NFD decomposition.
        var decomposed = input!.Normalize(NormalizationForm.FormD);

        // Strip combining marks.
        var stripped = new StringBuilder(decomposed.Length);
        foreach (var ch in decomposed)
        {
            if (CharUnicodeInfo.GetUnicodeCategory(ch) != UnicodeCategory.NonSpacingMark)
                stripped.Append(ch);
        }

        // Ampersand-token substitution (spaced form only).
        var ampersandSwapped = stripped
            .Replace(_AMPERSAND_TOKEN, _AND_TOKEN)
            .ToString();

        // Invariant-culture casefold.
        var lowered = CultureInfo.InvariantCulture.TextInfo.ToLower(ampersandSwapped);

        // Trim + collapse internal whitespace.
        return CollapseWhitespace(lowered.AsSpan().Trim());
    }

    private static string CollapseWhitespace(ReadOnlySpan<char> trimmed)
    {
        if (trimmed.IsEmpty)
            return string.Empty;

        var builder = new StringBuilder(trimmed.Length);
        var previousWasSpace = false;
        foreach (var ch in trimmed)
        {
            if (char.IsWhiteSpace(ch))
            {
                if (!previousWasSpace)
                {
                    builder.Append(_SPACE);
                    previousWasSpace = true;
                }
            }
            else
            {
                builder.Append(ch);
                previousWasSpace = false;
            }
        }

        return builder.ToString();
    }
}
